### DISCUSSION TITLE
set 'tee'

### DISCUSSION THREAD

**Discussion by 0ion9** [fish-shell/fish-shell#8885] [Ideas]
Personally I have observed that I often want to store the output of a complex command, while also immediately rendering it inline. Hence the 'tee' analogy. A toy example would be `touch /tmp/(set id (seq 100 | fzf); echo $id)`, allowing me to use $id in a subsequent command. Writing it out like this is clearly not DRY.

I've attempted to implement this as a function, but the main problem is that it seems impossible to `set -f` using that, as the result would be scoped to the helper function, not the function calling it.

My implementation follows:

`function settee;set -g $argv[1] $argv[2..]; printf %s\n $argv[2..];end`

used like this: `touch /tmp/(settee id (seq 10 | fzf).txt`

It could be improved to a) be more tee-like by piping in the input items (`seq 10 | fzf | settee id`, b) support -U, -x, -u, -a and -p. 

My question is, is there some ordinary way I'm missing to make `-f` scope support possible -- without outright patching `set` itself?

**Comment by 0ion9** [comment]
It seems `function [...] --no-scope-shadowing` may solve this for `-f`; Is there then a way to solve for `-l`? Less critical, since often -f will do instead.


ps. I guess arguably the function as proposed is really more like a `tee` version of `read` rather than of `set`.

**Reply by faho** [reply]
>Is there then a way to solve for -l

No. You're creating a function, so local scope will always get in the way. "local" is actually block-scoped and even with --no-scope-shadowing the new function still counts as a new block. It'll also still have issues with $argv because that's still shadowed even with --no-scope-shadowing.

(tbh --no-scope-shadowing typically isn't a good idea)

It would probably be okay to add a flag to `read` that both sets *and* prints the values. It already prints them if not given a variable name.

(but we should be careful with `read` because it's already too complicated and has annoying elements)
